coderd/wsbuilder/wsbuilder.go

```diff
@@ -0,0 +1,741 @@
+	ctx   context.Context
+	store database.Store
+
+	// cache of objects, so we only fetch once
+	template                  *database.Template
+	templateVersion           *database.TemplateVersion
+	templateVersionJob        *database.ProvisionerJob
+	templateVersionParameters *[]database.TemplateVersionParameter
+	lastBuild                 *database.WorkspaceBuild
+	lastBuildErr              *error
+	lastBuildParameters       *[]database.WorkspaceBuildParameter
+	lastParameterValues       *[]database.ParameterValue
+	lastBuildJob              *database.ProvisionerJob
+}
+
+type versionTarget struct {
+	active   bool
+	specific *uuid.UUID
+}
+
+type stateTarget struct {
+	orphan   bool
+	explicit *[]byte
+}
+
+func New(w database.Workspace, t database.WorkspaceTransition) Builder {
+	return Builder{workspace: w, trans: t}
+}
+
+// Methods that customize the build are public, have a struct receiver and return a new Builder.
+
+func (b Builder) VersionID(v uuid.UUID) Builder {
+	// nolint: revive
+	b.version = versionTarget{specific: &v}
+	return b
+}
+
+func (b Builder) ActiveVersion() Builder {
+	// nolint: revive
+	b.version = versionTarget{active: true}
+	return b
+}
+
+func (b Builder) State(state []byte) Builder {
+	// nolint: revive
+	b.state = stateTarget{explicit: &state}
+	return b
+}
+
+func (b Builder) Orphan() Builder {
+	// nolint: revive
+	b.state = stateTarget{orphan: true}
+	return b
+}
+
+func (b Builder) LogLevel(l string) Builder {
+	// nolint: revive
+	b.logLevel = l
+	return b
+}
+
+func (b Builder) Initiator(u uuid.UUID) Builder {
+	// nolint: revive
+	b.initiator = u
+	return b
+}
+
+func (b Builder) Reason(r database.BuildReason) Builder {
+	// nolint: revive
+	b.reason = r
+	return b
+}
+
+func (b Builder) LegacyParameterValues(p []codersdk.CreateParameterRequest) Builder {
+	// nolint: revive
+	b.legacyParameterValues = p
+	return b
+}
+
+func (b Builder) RichParameterValues(p []codersdk.WorkspaceBuildParameter) Builder {
+	// nolint: revive
+	b.richParameterValues = p
+	return b
+}
+
+// SetLastWorkspaceBuildInTx prepopulates the Builder's cache with the last workspace build.  This allows us
+// to avoid a repeated database query when the Builder's caller also needs the workspace build, e.g. auto-start &
+// auto-stop.
+//
+// CAUTION: only call this method from within a database transaction with RepeatableRead isolation.  This transaction
+// MUST be the database.Store you call Build() with.
+func (b Builder) SetLastWorkspaceBuildInTx(build *database.WorkspaceBuild) Builder {
+	// nolint: revive
+	b.lastBuild = build
+	return b
+}
+
+// SetLastWorkspaceBuildJobInTx prepopulates the Builder's cache with the last workspace build job.  This allows us
+// to avoid a repeated database query when the Builder's caller also needs the workspace build job, e.g. auto-start &
+// auto-stop.
+//
+// CAUTION: only call this method from within a database transaction with RepeatableRead isolation.  This transaction
```

Comment:
I don't think we expose that in our Store API.

It might be possible to extend it so you can get some transaction identifier, but that seems overkill for this use case.